server/src/main/java/org/apache/kafka/server/DefaultAutoTopicCreationManager.java

```diff
@@ -0,0 +1,255 @@
+import org.apache.kafka.common.requests.RequestContext;
+import org.apache.kafka.common.requests.RequestHeader;
+import org.apache.kafka.coordinator.group.GroupCoordinatorConfig;
+import org.apache.kafka.coordinator.share.ShareCoordinatorConfig;
+import org.apache.kafka.coordinator.transaction.TransactionLogConfig;
+import org.apache.kafka.server.common.ControllerRequestCompletionHandler;
+import org.apache.kafka.server.common.NodeToControllerChannelManager;
+import org.apache.kafka.server.config.AbstractKafkaConfig;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
+public class DefaultAutoTopicCreationManager implements AutoTopicCreationManager {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DefaultAutoTopicCreationManager.class);
+
+    private final AbstractKafkaConfig config;
+    private final NodeToControllerChannelManager channelManager;
+    private final Supplier<Properties> groupCoordinator;
+    private final Supplier<Properties> shareCoordinator;
+    private final Supplier<Properties> transactionTopicConfigsSupplier;
+    private final Set<String> inflightTopics = ConcurrentHashMap.newKeySet();
+
+    public DefaultAutoTopicCreationManager(
+            AbstractKafkaConfig config,
+            NodeToControllerChannelManager channelManager,
+            Supplier<Properties> groupCoordinatorConfigsSupplier,
+            Supplier<Properties> transactionTopicConfigsSupplier,
+            Supplier<Properties> shareCoordinatorConfigsSupplier
+    ) {
+        this.config = config;
+        this.channelManager = channelManager;
+        this.groupCoordinator = groupCoordinatorConfigsSupplier;
+        this.shareCoordinator = shareCoordinatorConfigsSupplier;
+        this.transactionTopicConfigsSupplier = transactionTopicConfigsSupplier;
+    }
+
+    @Override
+    public List<MetadataResponseTopic> createTopics(Set<String> topics, Optional<RequestContext> metadataRequestContext) {
+        var creatableTopics = new HashMap<String, CreatableTopic>();
+        var uncreatableTopicResponses = new ArrayList<MetadataResponseTopic>();
+        topics.forEach(topic -> {
+            // Attempt basic topic validation before sending any requests to the controller.
+            Optional<Errors> validationError;
+            if (!isValidTopicName(topic)) {
+                validationError = Optional.of(Errors.INVALID_TOPIC_EXCEPTION);
+            } else if (!inflightTopics.add(topic)) {
+                validationError = Optional.of(Errors.UNKNOWN_TOPIC_OR_PARTITION);
+            } else {
+                validationError = Optional.empty();
+            }
```

Comment:
nit: Could we use `Optional.empty` as a default value?
```suggestion
            Optional<Errors> validationError = Optional.empty();
            if (!isValidTopicName(topic)) {
                validationError = Optional.of(Errors.INVALID_TOPIC_EXCEPTION);
            } else if (!inflightTopics.add(topic)) {
                validationError = Optional.of(Errors.UNKNOWN_TOPIC_OR_PARTITION);
            }
```